# **1. DATA EXPLORATION**

In this notebook, we're gonna explore the data. At first, it is important to study the dataset and understand it because then we'll be able to focus on which areas of dataset to focus on more in future steps of the project. We'll only study the dataset as of now, all changes will be made in further notebooks

#### **1.1 Importing Libraries**

The following libraries are gonna be needed for this notebook:

1. Pandas
2. Numpy

Now we're gonna import them

In [1]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings("ignore")

#### **1.2 Loading Dataset**

We will now proceed to load the dataset. The dataset has been downloaded from Kaggle, the link for which is already in the Project Introduction notebook

In [2]:
df = pd.read_csv("../data/fraud_oracle.csv")
df.head()

,Month,WeekOfMonth,DayOfWeek,Make,AccidentArea,DayOfWeekClaimed,MonthClaimed,WeekOfMonthClaimed,Sex,MaritalStatus,...,AgeOfVehicle,AgeOfPolicyHolder,PoliceReportFiled,WitnessPresent,AgentType,NumberOfSuppliments,AddressChange_Claim,NumberOfCars,Year,BasePolicy
0,Dec,5,Wednesday,Honda,Urban,Tuesday,Jan,1,Female,Single,...,3 years,26 to 30,No,No,External,none,1 year,3 to 4,1994,Liability
1,Jan,3,Wednesday,Honda,Urban,Monday,Jan,4,Male,Single,...,6 years,31 to 35,Yes,No,External,none,no change,1 vehicle,1994,Collision
2,Oct,5,Friday,Honda,Urban,Thursday,Nov,2,Male,Married,...,7 years,41 to 50,No,No,External,none,no change,1 vehicle,1994,Collision
3,Jun,2,Saturday,Toyota,Rural,Friday,Jul,1,Male,Married,...,more than 7,51 to 65,Yes,No,External,more than 5,no change,1 vehicle,1994,Liability
4,Jan,5,Monday,Honda,Urban,Tuesday,Feb,2,Female,Single,...,5 years,31 to 35,No,No,External,none,no change,1 vehicle,1994,Collision


#### **1.3 Exploring The Dataset**

In this step, we're gonna explore the dataset. We'll perform some functions to understand more about the data like its shape, information, columns, etc.

##### **1.3.1 Dataset Shape**

In [3]:
df.shape

(15420, 33)

As we can see, the dataset has 15420 rows (or entries) and 33 columns (or features)

##### **1.3.2 Statistical Info**

The statistical information about the dataset will tell us all info about the numerical columns like total count, mean, minimum value and maximum value. This will help us know summary of numerical features

In [4]:
df.describe()

,WeekOfMonth,WeekOfMonthClaimed,Age,FraudFound_P,PolicyNumber,RepNumber,Deductible,DriverRating,Year
count,15420.000000,15420.000000,15420.000000,15420.000000,15420.000000,15420.000000,15420.000000,15420.000000,15420.000000
mean,2.788586,2.693969,39.855707,0.059857,7710.500000,8.483268,407.704280,2.487808,1994.866472
std,1.287585,1.259115,13.492377,0.237230,4451.514911,4.599948,43.950998,1.119453,0.803313
min,1.000000,1.000000,0.000000,0.000000,1.000000,1.000000,300.000000,1.000000,1994.000000
25%,2.000000,2.000000,31.000000,0.000000,3855.750000,5.000000,400.000000,1.000000,1994.000000
50%,3.000000,3.000000,38.000000,0.000000,7710.500000,8.000000,400.000000,2.000000,1995.000000
75%,4.000000,4.000000,48.000000,0.000000,11565.250000,12.000000,400.000000,3.000000,1996.000000
max,5.000000,5.000000,80.000000,1.000000,15420.000000,16.000000,700.000000,4.000000,1996.000000


##### **1.3.3 Dataset Info**

The info of dataset will tell us the non-null count, datatype and name of each column. This will help us understand if there is any need for modifying the datatype of any column

In [5]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 15420 entries, 0 to 15419
Data columns (total 33 columns):
 #   Column                Non-Null Count  Dtype
---  ------                --------------  -----
 0   Month                 15420 non-null  str  
 1   WeekOfMonth           15420 non-null  int64
 2   DayOfWeek             15420 non-null  str  
 3   Make                  15420 non-null  str  
 4   AccidentArea          15420 non-null  str  
 5   DayOfWeekClaimed      15420 non-null  str  
 6   MonthClaimed          15420 non-null  str  
 7   WeekOfMonthClaimed    15420 non-null  int64
 8   Sex                   15420 non-null  str  
 9   MaritalStatus         15420 non-null  str  
 10  Age                   15420 non-null  int64
 11  Fault                 15420 non-null  str  
 12  PolicyType            15420 non-null  str  
 13  VehicleCategory       15420 non-null  str  
 14  VehiclePrice          15420 non-null  str  
 15  FraudFound_P          15420 non-null  int64
 16  PolicyNumber   

##### **1.3.4 Checking Target Variable**

Now we're gonna have a look at our main target feature which is `FraudFound_P`. From this, we'll understand how out target feature works and the total class imbalance 

In [6]:
df["FraudFound_P"].value_counts()

FraudFound_P
0    14497
1      923
Name: count, dtype: int64

In [7]:
df["FraudFound_P"].value_counts(normalize=True)*100

FraudFound_P
0    94.014267
1     5.985733
Name: proportion, dtype: float64

As we can see, the non-fraud values are 94.01% and fraud values are 5.98%. This shows a big class imbalance. This can cause some issues in our overall project so we'll have to address it properly in future notebooks

##### **1.3.5 Checking Missing Values**

Here, we'll check if our dataset contains any missing values and deal with it appropriately

In [8]:
df.isna().sum().sum()

np.int64(0)

As we can see, there are no null values in our dataset. So we don't need to handle them

##### **1.3.6 Duplicate Values**

Here, we'll check if our dataset contains any duplicate rows. This checks rows where all columns are exactly the same

In [9]:
df.duplicated().sum()

np.int64(0)

As we can see, there are no duplicate values in our dataset. So again, no need to handle them. We can notice that there are no missing or duplicate values in our dataset, which is completely clean to work with

#### **1.4 Checking Columns**

In this step, we're gonna check all the columns of our dataset. We'll check different columns to see how to deal with them

##### **1.4.1 All Columns**

First we'll have a look at all the columns in our dataset

In [10]:
df.columns

Index(['Month', 'WeekOfMonth', 'DayOfWeek', 'Make', 'AccidentArea',
       'DayOfWeekClaimed', 'MonthClaimed', 'WeekOfMonthClaimed', 'Sex',
       'MaritalStatus', 'Age', 'Fault', 'PolicyType', 'VehicleCategory',
       'VehiclePrice', 'FraudFound_P', 'PolicyNumber', 'RepNumber',
       'Deductible', 'DriverRating', 'Days_Policy_Accident',
       'Days_Policy_Claim', 'PastNumberOfClaims', 'AgeOfVehicle',
       'AgeOfPolicyHolder', 'PoliceReportFiled', 'WitnessPresent', 'AgentType',
       'NumberOfSuppliments', 'AddressChange_Claim', 'NumberOfCars', 'Year',
       'BasePolicy'],
      dtype='str')

##### **1.4.2 Numerical Columns**

Now we'll have a look at the numerical only columns in our dataset

In [11]:
df.select_dtypes(include=np.number).columns

Index(['WeekOfMonth', 'WeekOfMonthClaimed', 'Age', 'FraudFound_P',
       'PolicyNumber', 'RepNumber', 'Deductible', 'DriverRating', 'Year'],
      dtype='str')

##### **1.4.3 Categorical Columns**

Now we'll have a look at the categorical only columns in our dataset. These are columns which contain string and other object values in them

In [12]:
df.select_dtypes(include=["object"]).columns

Index(['Month', 'DayOfWeek', 'Make', 'AccidentArea', 'DayOfWeekClaimed',
       'MonthClaimed', 'Sex', 'MaritalStatus', 'Fault', 'PolicyType',
       'VehicleCategory', 'VehiclePrice', 'Days_Policy_Accident',
       'Days_Policy_Claim', 'PastNumberOfClaims', 'AgeOfVehicle',
       'AgeOfPolicyHolder', 'PoliceReportFiled', 'WitnessPresent', 'AgentType',
       'NumberOfSuppliments', 'AddressChange_Claim', 'NumberOfCars',
       'BasePolicy'],
      dtype='str')

##### **1.4.4 Binary Columns**

Now we'll check the binary columns. These are columns which have only 2 categories (which can be converted to either yes/no or 0/1)

In [13]:
binary_cols = [col for col in df.columns if df[col].nunique() == 2]
binary_cols

['AccidentArea',
 'Sex',
 'Fault',
 'FraudFound_P',
 'PoliceReportFiled',
 'WitnessPresent',
 'AgentType']

#### **1.5 Notebook Summary**

**Dataset Exploration:**

- Dataset shape is 15420x33
- Target value distribution is ~94% non-fraud and ~6% fraud
- Checked numerical, categorical and binary columns

**Insights:**

- We have to manage class imbalance
- We have to check some columns which have wrong datatypes